### PR TITLE
Refactor NVidia drivers

### PR DESCRIPTION
**Pull request by abbradar**
###### Motivation for this change
 
Make several improvements to nvidia drivers on NixOS:

* Use libglvnd;
* Unify nearly all derivations and builders (leave only legacy_174 -- it's too much different);
* Build more driver parts from source;
* Default to GTK3 (#18559).

###### Things done

- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

Sadly I can't fully test this myself because at some point my laptop lost ability to switch to dedicated-only card in BIOS so I can't use nvidia without intel/bumblebee now. Bumblebee works fine though, both `optirun` and `primusrun`.

### CODE REVIEW

**Comment by grahamc** [comment]
What laptop do you have? ps: if you rebase this against master, https://prs.nix.gsc.io/eval/857 will build the nixos/release.nix + tests.

**Comment by abbradar** [comment]
@grahamc Lenovo T440p. I think at some point option to force NVidia only was there, but it's gone now (I regularly update BIOS so that may be the reasion).

Would Hydra build unfree packages and their reverse dependencies? Also, will use that chance to say thank you very much for your work on Hydra for PRs ^_^.

**Comment by grahamc** [comment]
Hmm... I wish it would. I'm working on having this included in hydra.nixos.org. Unfortunately, I don't want to set a precedent that it'll build unfree :(

**Comment by grahamc** [comment]
Also, you're welcome, I'm glad to make progress on this! :)

**Comment by abbradar** [comment]
> Unfortunately, I don't want to set a precedent that it'll build unfree :(
That's completely fine, on the contrary -- I'd be surprised if you did enable this :D

File: pkgs/os-specific/linux/nvidia-x11/generic.nix [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        fetchurl {
+          url = "http://download.nvidia.com/XFree86/Linux-x86_64/${version}/NVIDIA-Linux-x86_64-${version}${pkgSuffix}.run";
+          sha256 = sha256_64bit;
+        }
+      else throw "nvidia-x11 does not support platform ${stdenv.system}";
+
+    inherit version useGLVND;
+    inherit (stdenv) system;
+
+    outputs = [ "out" ] ++ optional (!libsOnly) "bin";
```
Comment (dezgeg) [inline pkgs/os-specific/linux/nvidia-x11/generic.nix]: The convention has been to put the output containing binaries as the first output (it changed at one point).

Also I didn't check too closely, but doesn't splitting to multiple outputs remove the need for all the `libsOnly` logic?

**Comment by abbradar** [comment]
Right, forgot about output changes. 

Outputs here solve the problem of space but not of time - without `libsOnly` users with multilib OpenGL would need to compile 32-bit nvidia kernel modules each update. 
-- 
Nikolay.

**Comment by dezgeg** [comment]
Hmm, what's the benefit of splitting this to multiple outputs then? In what situations you end up with the built system depending on only one of the outputs?

**Comment by abbradar** [comment]
@dezgeg Never ;) The point of splitting is to keep logically separate parts separate. This way we don't have kernel modules and OpenGL libraries in system environment, for example.

BTW is there a way to enforce current convention (binary outputs primary) when the binary output doesn't always exist?

**Comment by vcunat** [comment]
I can test on a notebook hard-wired to nvidia.

**Comment by abbradar** [comment]
@vcunat would be cool!

BTW (offtopic, teaser) I have an in-progress branch that moves us to libglvnd completely (so `mesa_noglu` is used only for drivers and `libOSMesa`). However I'll hold it until some other distro makes this move and irons the wrinkles (IIRC Fedora already started) -- Mesa still doesn't support EGL and GLES via libglvnd, and even GLX is still very fresh.

Hopefully when AMDGPU-PRO also moves to `libglvnd` we can finally get rid of `LD_LIBRARY_PATH`!

**Comment by vcunat** [comment]
@abbradar: [my older libglvnd WIP](https://github.com/vcunat/nixpkgs/commits/p/libGL) if you find some useful bits... but I expect most work will be in finishing the details, especially if we were to implement putting the C symbols into a separate linker namespace.

**Comment by vcunat** [comment]
I originally paused the effort for AMD to catch up; that might be a long wait, so waiting for at least some larger distro might save some work.

**Comment by abbradar** [comment]
Yep, I think there are also more drivers that won't support `libglvnd` any time soon (maybe VMware/some other proprietary VMs?) so we can't drop `LD_LIBRARY_PATH` support altogether, but waiting for Mesa and some distro to make the move first seems a good idea. Then we can at least link everything to `libglvnd`.

**Comment by vcunat** [comment]
I think we can link anyway during build-time and then just put another libGL on `$LD_LIBRARY_PATH` instead, if the driver requires it.

**Comment by vcunat** [comment]
I'm specifically looking forward to mesa updates not causing a mass rebuilds.

**Comment by vcunat** [comment]
Driver seems OK, but `nvidia-settings` can't find any GTK.

**Comment by vcunat** [comment]
The `libnvidia-gtk3.so` that gets to `/run/opengl-driver/lib` can't find most of libs.  Strangely there's some such library in `nvidia-settings-375.26/lib` also but it's got good linkage but it's missing version symlinks and executable bit.

**Comment by abbradar** [comment]
Ah, during some round of final refactoring I lost `rm libnvidia-gtk*` (we build it from source now). Try again please!

**Comment by abbradar** [comment]
Resolved conflicts.

**Comment by vcunat** [comment]
The settings app does work now for me, and the rest has been working OK the whole time - now even with xorg-server-1.19.1.  I don't use any special stuff like vulkan or CUDA.  Anything else to wait for?  It doesn't seem likely to get much more testing until it's forced on unstable/master users :-)

**Comment by abbradar** [comment]
Actually I was just waiting for you to confirm that settings now work -- we got a deadlock :D

**Comment by vcunat** [comment]
JFTR, some of this bunch of lines from journalctl seem fishy:
```
kernel: [drm] [nvidia-drm] [GPU ID 0x00000100] Loading driver
systemd-modules-load[509]: Inserted module 'nvidia_uvm'
systemd-udevd[591]: Process '/nix/store/b3hnjqy2py1hbh364i6jj6jmh9ajm79l-bash-4.4-p5/bin/bash -c 'mknod -m 666 /dev/nvidia-uvm c $(grep nvidia-uvm /proc/devices | cut -d \  -f 1) 0'' failed with exit code 1.
kernel: nvidia-modeset: Allocated GPU:0 (GPU-4e7b3168-e136-eb9b-a86c-613b8e52234e) @ PCI:0000:01:00.0
nvidia-persistenced[3690]: Failed to create directory /var/run/nvidia-persistenced: Permission denied
nvidia-persistenced[3690]: Unable to access /var/run/nvidia-persistenced: No such file or directory
nvidia-persistenced[3690]: Shutdown (3690)
nvidia-settings[3713]: Error retrieving accessibility bus address: org.freedesktop.DBus.Error.ServiceUnknown: The name org.a11y.Bus was not provided by any .service files
xsession[1274]: ERROR: nvidia-settings could not find the registry key file. This file
xsession[1274]:        /usr/share/nvidia/nvidia-application-profiles-375.26-key-documentati
xsession[1274]:        /usr/share/nvidia/nvidia-application-profiles-key-documentation. The
nvidia-settings[3713]: Could not load a pixbuf from /org/gtk/libgtk/theme/Adwaita/assets/check-symbolic.svg.
nvidia-settings[3713]: Allocating size to CtkWindow 0xb462a0 without calling gtk_widget_get_preferred_width/height(). How does the code know the size to allocate?
nvidia-settings[4815]: Error retrieving accessibility bus address: org.freedesktop.DBus.Error.ServiceUnknown: The name org.a11y.Bus was not provided by any .service files
xsession[1274]: ERROR: nvidia-settings could not find the registry key file. This file
xsession[1274]:        /usr/share/nvidia/nvidia-application-profiles-375.26-key-documentati
xsession[1274]:        /usr/share/nvidia/nvidia-application-profiles-key-documentation. The
nvidia-settings[4815]: Could not load a pixbuf from /org/gtk/libgtk/theme/Adwaita/assets/check-symbolic.svg.
nvidia-settings[4815]: Allocating size to CtkWindow 0x24922a0 without calling gtk_widget_get_preferred_width/height(). How does the code know the size to allocate?

```

**Comment by abbradar** [comment]
@vcunat Those look interesting:

1) I'm confused at what tries to run `nvidia-persistenced`. I have packaged it but without a NixOS module and such -- it is supposed to be a daemon which keeps CUDA state in memory or something like that;
2) Paths for application profiles need to be fixed, I'll look into that.

**Comment by abbradar** [comment]
Can you track at which moment `nvidia-persistenced` is run? I.e. during start of user session, when you start settings, during boot...

**Comment by vcunat** [comment]
I think I invoked it from my user session by a mistake.

**Comment by vcunat** [comment]
Yes, logs confirm it wasn't done during startup.

Now I also noticed:
```
xsession[1274]: Crash Annotation GraphicsCriticalError: |[0][GFX1-]: GLContext is disabled due to a previous crash.|.... (repeats of all the same)
```

**Comment by abbradar** [comment]
Ugh, any idea what process does this? It seems it's dying in a loop, so I'd be difficult to lookup it by pid. 
-- 
Nikolay.

**Comment by vcunat** [comment]
No idea yet.  After switching xorg-server to 1.19.1, my firefox shows github's top bar with black background instead of light, though that might be unrelated.  I keep watching the system, but so far I've felt no problems.

**Comment by abbradar** [comment]
I have the black bar in qutebrowser, Chromium and Firefox since yesterday. Seems not related to nvidia at least, though strange. 
-- 
Nikolay.

**Comment by grahamc** [comment]
The black bar has been present in GitHub Enterprise for months, apparently they've moved it over to public GitHub. FWIW.

**Comment by abbradar** [comment]
Oh, that explains it, thanks.

**Comment by vcunat** [comment]
:-D  I tried on chromium and didn't have the bar.  I assumed won't change it, but I was wrong.

**Comment by dezgeg** [comment]
https://github.com/DennisSnijder/MakeGithubGreatAgain

**Pull request merged**